### PR TITLE
React to NATS backend config changes

### PR DESCRIPTION
**Pull request by grischperl**
**Description**

Changes proposed in this pull request:

- Add config hash to `EventingStatus`
- React to changes in backend config (so far to all changes, needs to be changed to NATS specific)
- Add tests

**Related issue(s)**

#45

### CODE REVIEW

File: internal/controller/eventing/nats.go [hunk 1]
```diff
@@ -66,7 +119,9 @@ func (r *Reconciler) stopNATSSubManager(runCleanup bool, log *zap.SugaredLogger)
 	log.Info("NATS subscription-manager stopped!")
 	// update flags so it does not try to stop the manager again.
 	r.isNATSSubManagerStarted = false
-	r.natsSubManager = nil
+	if runCleanup {
```
Comment (muralov) [inline internal/controller/eventing/nats.go:122]: why don't you want to keep old manager if clean-up is not wished?

**Review by muralov — CHANGES_REQUESTED** [review]

File: internal/controller/eventing/nats_test.go [hunk 2]
```diff
@@ -138,6 +145,33 @@ func Test_reconcileNATSSubManager(t *testing.T) {
 			wantAssertCheck:  true,
 			givenShouldRetry: true,
 			wantError:        errors.New("failed to start"),
+			wantHashAfter:    int64(6872426080962807353),
```
[the remark this answers is not in the document]
Comment (grischperl) [reply internal/controller/eventing/nats_test.go:148]: Done!

File: internal/controller/eventing/nats.go [hunk 3]
```diff
@@ -29,28 +40,59 @@ func (r *Reconciler) reconcileNATSSubManager(eventing *v1alpha1.Eventing, log *z
...
+			}
+
+			// stop the subsManager without cleanup
+			if r.isNATSSubManagerStarted {
+				if err := r.stopNATSSubManager(false, log); err != nil {
+					return err
+				}
+				return nil
+			}
+		}
```
Comment (muralov) [inline internal/controller/eventing/nats.go:60]: you may take this out of else block, what do you think? It makes more readable, nested ifs are not so readable.

File: internal/controller/eventing/nats_test.go [hunk 4]
```diff
@@ -138,17 +162,47 @@ func Test_reconcileNATSSubManager(t *testing.T) {
 			wantAssertCheck:  true,
 			givenShouldRetry: true,
 			wantError:        errors.New("failed to start"),
+			wantHashAfter:    uint64(0),
+		},
+		{
+			name:                         "it should update the subscription manager when the backend config changes",
+			givenIsNATSSubManagerStarted: false,
+			givenHashBefore:              uint64(0),
```
Comment (muralov) [inline internal/controller/eventing/nats_test.go:170]: should it not have a hash before to see if existing has changes to a new one?

File: internal/controller/eventing/nats.go (hunk 3 above)
Comment (grischperl) [reply internal/controller/eventing/nats.go]: Done

File: internal/controller/eventing/nats_test.go (hunk 4 above)
Comment (grischperl) [reply internal/controller/eventing/nats_test.go]: The test used to go through the whole flow: empty subManager -> gets a hash -> then update the manager -> new hash. But as the case of the new subManager is already covered, I changed this to not empty subManager -> update config -> new hash.

**Review by muralov — APPROVED** [review]

**Review by muralov — CHANGES_REQUESTED** [review]
please fix unit test.

**Review by muralov — APPROVED** [review]

**Pull request merged**